packages/core/contracts/oracle/implementation/cross-chain/BeaconOracle.sol

```diff
@@ -0,0 +1,150 @@
+        Publisher // Can publish price request results.
+    }
+
+    struct Price {
+        bool isAvailable;
+        int256 price;
+    }
+
+    // The two structs below are used in an array and mapping to keep track of prices that have been requested but are
+    // not yet available.
+    struct QueryIndex {
+        bool isValid;
+        uint256 index;
+    }
+
+    // Represents a (identifier, time) point that has been queried.
+    struct QueryPoint {
+        bytes32 identifier;
+        uint256 time;
+        bytes ancillaryData;
+    }
+
+    // Conceptually we want a (time, identifier, ancillaryData) -> price map.
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => Price))) internal verifiedPrices;
+
+    // The mapping and array allow retrieving all the elements in a mapping and finding/deleting elements.
+    // Can we generalize this data structure?
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => QueryIndex))) internal queryIndices;
+
+    QueryPoint[] public requestedPrices;
+
+    // Finder to provide addresses for DVM contracts.
+    FinderInterface public finder;
+
+    event PriceRequestAdded(address indexed requester, bytes32 indexed identifier, uint256 time, bytes ancillaryData);
+    event PushedPrice(
+        address indexed pusher,
+        bytes32 indexed identifier,
+        uint256 time,
+        bytes ancillaryData,
+        int256 price
+    );
+
+    /**
+     * @notice Constructor.
+     * @param _finderAddress finder to use to get addresses of DVM contracts.
+     */
+    constructor(address _finderAddress) public {
+        finder = FinderInterface(_finderAddress);
+    }
+
+    /**
+     * @notice Enqueues a request (if a request isn't already present) for the given (identifier, time, ancillary data)
+     * pair.
+     * @dev If this is a SinkOracle, then this should be called by an OptimisticOracle on same network that wants to
+     * bubble up price request to the Mainnet DVM. If this is a SourceOracle, it will pass on the request to the DVM.
+     */
+    function _requestPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) internal onlyRoleHolder(uint256(Roles.Requester)) {
+        Price storage lookup = verifiedPrices[identifier][time][ancillaryData];
+        if (!lookup.isAvailable && !queryIndices[identifier][time][ancillaryData].isValid) {
+            // New query, enqueue it for review.
+            queryIndices[identifier][time][ancillaryData] = QueryIndex(true, requestedPrices.length);
+            requestedPrices.push(QueryPoint(identifier, time, ancillaryData));
+            emit PriceRequestAdded(msg.sender, identifier, time, ancillaryData);
+        }
+    }
+
+    /**
+     * @notice Pushes the verified price for a requested query.
+     * @dev Should be called by an off-chain relayer who saw a price resolved on a different network's Source or Sink
+     * Oracle.
+     */
+    function _pushPrice(
```

Comment:
A lot of this functionality seems like it was pulled from the mock oracle, but I think we can simplify if we don't need to have a method that returns all of the unresolved price requests (which we do have in the mock oracle, but don't seem to need here).

If you agree, I think we could drop the 3 data structures and just have a single mapping from the (identifier, timestamp, ancillary data) to a struct that stores:
1. The state of the request: (NeverRequested, Requested, Resolved).
2. The price.

This would also mean these methods get a lot simpler.

For request price, all we would need to do is check the current state of the request and update it.

For push price, we would just check the state and set the price.

I think both would end up being 4-ish lines if I'm thinking about it correctly.

Thoughts?

Reply:
I agree let's simplify. I do think we should be able to return pending price requests for convenience however, so perhaps we make the price requests struct pblic?

Reply:
I would argue that since we don't need it on-chain, we could just use events to construct it. If we do want to be able to return a list, the data structures as you have them may make sense -- I think that list is the source of the complexity if I remember correctly.

Reply:
Yeah removing that list simplies things a lot

Reply:
+1